boards: Add support for the iMX6SX Udoo Neo board

MaureenHelm opened the pull request:
Adds support for the iMX6SX M4 subsystem on the Udoo Neo board. This initial port includes uart and gpio drivers to enable the hello_world and basic/blinky samples.

cc: @diegosueiro

stanislav-poboril committed:
ext/hal/nxp/imx: Import the nxp imx6 freertos bsp

This code component is used to add Zephyr support on iMX6SX
processors, exclusively on Cortex M4 core, and to speed up the
development process it was decided to have it based on NXP FreeRTOS BSP
implementation.

The i.MX FreeRTOS BSP is split into separate downloadable packages,
based on SoC. The packages share most of the peripheral driver files
and here they are combined together.

The source code was imported from the following folders:
FreeRTOS_BSP_1.0.1_iMX6SX/platform/drivers
FreeRTOS_BSP_1.0.1_iMX6SX/platform/devices

This source code depends on headers and sources from zephyr:
    ext/hal/cmsis

Origin: i.MX 6SoloX FreeRTOS BSP 1.0.1 for Cortex-M4 Peripheral Driver
License: BSD 3-Clause
URL: https://www.nxp.com/webapp/Download?colCode=FreeRTOS_MX6SX_1.0.1_LINUX&appType=license
commit: no commit hash
Purpose: The peripheral driver wraps the H/W for i.MX6SX M4 core
Maintained-by: External

Signed-off-by: Stanislav Poboril <stanislav.poboril@nxp.com>

stanislav-poboril committed:
arch: Add imx6sx m4 soc support

The i.MX 6SoloX SoC is a hybrid multi-core processor composed by one
Cortex A9 core and one Cortex M4 core.

Zephyr was ported to run on the M4 core. In a later release, it will
also communicate with the A9 core (running Linux) via RPMsg.

The low level drivers come from NXP FreeRTOS BSP and are located at
ext/hal/nxp/imx. More details can be found at ext/hal/nxp/imx/README

The A9 core is responsible to load the M4 binary application into the
RAM, put the M4 in reset, set the M4 Program Counter and Stack Pointer,
and get the M4 out of reset.
The A9 can perform these steps at bootloader level after the Linux
system has booted.

Signed-off-by: Stanislav Poboril <stanislav.poboril@nxp.com>

stanislav-poboril committed:
boards/arm: add support for udoo_neo_full_m4 board

The UDOO Neo Full single board computer configuration supports
the following hardware features on the Cortex M4 Core:

+-----------+------------+-------------------------------------+
| Interface | Controller | Driver/Component                    |
+===========+============+=====================================+
| NVIC      | on-chip    | nested vector interrupt controller  |
+-----------+------------+-------------------------------------+
| SYSTICK   | on-chip    | systick                             |
+-----------+------------+-------------------------------------+
| UART      | on-chip    | serial port-polling;                |
|           |            | serial port-interrupt               |
+-----------+------------+-------------------------------------+
| GPIO      | on-chip    | general purpose input/output        |
+-----------+------------+-------------------------------------+

The default configuration can be found in the defconfig file:
boards/arm/udoo_neo_full_m4/udoo_neo_full_m4_defconfig

Other hardware features are not currently supported by the port.

Connections and IOs:

The UDOO Neo Full board was tested with the following pinmux
controller configuration.

+---------------+-----------------+---------------------------+
| Board Name    | SoC Name        | Usage                     |
+===============+=================+===========================+
| J4 RX         | UART5_RX_DATA   | UART Console              |
+---------------+-----------------+---------------------------+
| J4 TX         | UART5_TX_DATA   | UART Console              |
+---------------+-----------------+---------------------------+

The board has been tested with the following samples.
 - samples/hello_world
 - samples/basic/blinky

Signed-off-by: Stanislav Poboril <stanislav.poboril@nxp.com>

dbkinder commented on boards/arm/udoo_neo_full_m4/doc/udoo_neo_full.rst:
```diff
@@ -0,0 +1,216 @@
+
+UDOO Neo Full is an open source Arduino Uno compatible single board computer.
+It is equipped with an NXP |reg| i.MX 6SoloX hybrid multicore processor
+composed by one ARM |reg| Cortex-A9 core running up to 1 GHz and one Cortex-M4
```
change to ``composed of``

dbkinder commented on boards/arm/udoo_neo_full_m4/doc/udoo_neo_full.rst:
```diff
@@ -0,0 +1,216 @@
+It is equipped with an NXP |reg| i.MX 6SoloX hybrid multicore processor
+composed by one ARM |reg| Cortex-A9 core running up to 1 GHz and one Cortex-M4
+core running up to 227 MHz for high CPU performance and real-time response.
+Zephyr was ported to run on the Cortex-M4 core. In a later release, it will
```
for clarity add ``only`` to:

   ported to run on the Cortex-M4 core only.

dbkinder commented on boards/arm/udoo_neo_full_m4/doc/udoo_neo_full.rst:
```diff
@@ -0,0 +1,216 @@
+Zephyr was ported to run on the Cortex-M4 core. In a later release, it will
+also communicate with the Cortex-A9 core (running Linux) via RPMsg.
+
+.. image:: udoo_neo_full.jpg
```
The ``image`` directive doesn't allow a caption (hence the error about content not being allowed.  Change this to use the ``figure`` directive (everything else is the same), which does allow content (for the caption).

dbkinder commented on boards/arm/udoo_neo_full_m4/doc/udoo_neo_full.rst:
```diff
@@ -0,0 +1,216 @@
+
+The MCIMX6X SoC is configured to use the 24 MHz external oscillator
+on the board with the on-chip PLL to generate core clock.
+PLL settings for M4 core are set via code running on A9 core.
```
change to:  ``on the A9 core``

dbkinder commented on boards/arm/udoo_neo_full_m4/doc/udoo_neo_full.rst:
```diff
@@ -0,0 +1,216 @@
+Serial Port
+===========
+
+The MCIMX6X SoC has six UARTs. UART5 is configured for M4 console and the
```
change to ``for the M4 core`` and ``by the A9 core`` on the next line

dbkinder commented on boards/arm/udoo_neo_full_m4/doc/udoo_neo_full.rst:
```diff
@@ -0,0 +1,216 @@
+at power-on-reset. Therefore it needs to be started by the A9 core.
+The A9 core is responsible to load the M4 binary application into the RAM,
+put the M4 in reset, set the M4 Program Counter and Stack Pointer, and get
+the M4 out of reset. The A9 can perform these steps at bootloader level
```
change to ``at the bootloader``

dbkinder commented on boards/arm/udoo_neo_full_m4/doc/udoo_neo_full.rst:
```diff
@@ -0,0 +1,216 @@
+Debugging
+=========
+
+The UDOO Neo Full board includes the pads for soldering of the 14-pin JTAG
```
change to:

    The UDOO Neo Full board includes pads for soldering the 14-pin JTAG
    connector. Zephyr applications running on the M4 core have only been
    tested by observing UART console output.

galak reviewed (commented):
From the perspective of Zephyr & the m4, does the the fact that the board is udoo_neo_full vs the other cfg mater?

galak commented on boards/arm/udoo_neo_full_m4/doc/udoo_neo_full.rst:
```diff
@@ -0,0 +1,216 @@
+Programming and Debugging
+*************************
+
+The M4 core does not have flash memory and is not provided a clock
```
Can we add details like in colibri_imx7d_m4 doc's have for loading via u-boot?

diegosueiro replied:
Agreed

stanislav-poboril replied:
I haven't used u-boot. Only loaded it from Linux user space application which is part of the particular Linux distribution used (UDOObuntu). Anyway, I will add some instructions for this distribution at least (and the link to program, which is distribution independent).

stanislav-poboril replied:
Added details about the loading of second core image.

diegosueiro commented on boards/arm/udoo_neo_full_m4/doc/udoo_neo_full.rst:
```diff
@@ -0,0 +1,216 @@
+The A9 core is responsible to load the M4 binary application into the RAM,
+put the M4 in reset, set the M4 Program Counter and Stack Pointer, and get
+the M4 out of reset. The A9 can perform these steps at bootloader level
+after the Linux system has booted.
```
Need to add an "or".
`at the bootloader level or after the Linux system has booted.`

diegosueiro commented on boards/arm/udoo_neo_full_m4/pinmux.c:
```diff
@@ -0,0 +1,52 @@
+#include <init.h>
+#include "device_imx.h"
+
+static int udoo_neo_full_m4_init(struct device *dev)
```
What about the GPIO pinmux config?

stanislav-poboril replied:
Added LED GPIO pin config into pinmux.c + RDC access in soc.c

diegosueiro commented on drivers/gpio/CMakeLists.txt:
```diff
@@ -7,6 +7,7 @@ zephyr_sources_ifdef(CONFIG_GPIO_DW         gpio_dw.c)
 zephyr_sources_ifdef(CONFIG_GPIO_ESP32      gpio_esp32.c)
 zephyr_sources_ifdef(CONFIG_GPIO_FE310      gpio_fe310.c)
 zephyr_sources_ifdef(CONFIG_GPIO_GECKO      gpio_gecko.c)
+zephyr_sources_ifdef(CONFIG_GPIO_IMX        gpio_imx_gpio.c)
```
Do you think is possible to use the same GPIO driver from PR #7113?

stanislav-poboril replied:
Yes, should be possible. We should select one or the other.

diegosueiro replied:
Instead of gpio_imx_gpio.c can we use gpio_imx.c?
This way the filename is more clear.

stanislav-poboril replied:
Ok. Note that the FreeRTOS BSP driver is also gpio_imx.c, hopefully it won't cause problems in any toolchains.

stanislav-poboril replied:
Extracted this part into a separated PR #7306 so #7113 can rebase to it.

diegosueiro replied:
@stanislav-poboril,

I had this name in my PR and no issues with the compilation.

diegosueiro commented on boards/arm/udoo_neo_full_m4/udoo_neo_full_m4_defconfig:
```diff
@@ -0,0 +1,19 @@
+CONFIG_SERIAL=y
+CONFIG_SERIAL_HAS_DRIVER=y
+CONFIG_CORTEX_M_SYSTICK=y
+CONFIG_GPIO=n
```
In the board dts you set the gpio ports 4, 5 and 6 to status "ok" but in this config the GPIO is set to "n".

stanislav-poboril replied:
Changed.

diegosueiro commented on arch/arm/soc/nxp_imx/mcimx6x_m4/soc.c:
```diff
@@ -0,0 +1,160 @@
+ *
+ * @return 0
+ */
+static int mcimx6x_m4_init(struct device *arg)
```
Should we have a similar soc initialization for nxp_mcimx7_init, like SOC_CacheInit()?

stanislav-poboril replied:
I think this is optional. The FreeRTOS BSP initializes it when-in startup, see:
FreeRTOS_BSP_1.0.1_iMX7D/platform/devices/MCIMX7D/startup/system_MCIMX7D_M4.c, function SystemInit(). Or we could have a config item at the SoC level to enable/disable this.
We could also utilize LMEM_EnableSystemCache() in ext/hal/nxp/imx/drivers/lmem.c.

diegosueiro commented on boards/arm/udoo_neo_full_m4/udoo_neo_full_m4.yaml:
```diff
@@ -0,0 +1,13 @@
+identifier: udoo_neo_full_m4
+name: UDOO Neo Full
+type: mcu
+arch: arm
```
Add the entry:
`ram: 32`

dbkinder commented on boards/arm/udoo_neo_full_m4/doc/udoo_neo_full.rst:
```diff
@@ -0,0 +1,216 @@
+
+References
+==========
+
```
If you're intention was for the references listed here to show up in the generated HTML page under this heading, add a directive to do that here:

    .. target-notes::

followed by a blank line.

stanislav-poboril commented:
> From the perspective of Zephyr & the m4, does the the fact that the board is udoo_neo_full vs the other cfg mater?

The basic, extended and full versions differ in the availability of:
* Wi-Fi
* Bluetooth
* Fast Ethernet
* 9-Axis sensor
* amount of external RAM (this depends on size of area reserved for M4 anyway)

So it makes sense to have separate dts files at least. Question is if Zephyr will support all those peripherals. If not, we don't have to distinguish versions. Or we could only separate the configuration once something from the above list is implemented.

diegosueiro commented on arch/arm/soc/nxp_imx/mcimx6x_m4/dts.fixup:
```diff
@@ -0,0 +1,98 @@
+
+#define CONFIG_UART_IMX_UART_1_BASE_ADDRESS	NXP_IMX_UART_42020000_BASE_ADDRESS_0
+#define CONFIG_UART_IMX_UART_1_NAME		NXP_IMX_UART_42020000_LABEL
+#define CONFIG_UART_IMX_UART_1_IRQ		NXP_IMX_UART_42020000_IRQ_0
```
Use `CONFIG_UART_IMX_UART_[X]_IRQ_NUM` instead of `CONFIG_UART_IMX_UART_[X]_IRQ`

stanislav-poboril replied:
Changed.

dbkinder reviewed (approved):
Thanks.

galak commented:
Can we update this to resolve the conflict?

stanislav-poboril commented:
> Can we update this to resolve the conflict?
Yes, I am planning to update my open pull requests, but I will get into this tomorrow or in the week of May 21st.

MaureenHelm commented:
- Rebased to latest master
- Fixed CI UART failures
- Updated dts base addresses to drop trailing _0 (consequence of commit 081c9c3)

I don't have this board on hand so I could not test these changes on hardware.

@galak: Are these the conflicts you were referring to, or is there something else? There weren't any git merge conflicts when I rebased.

stanislav-poboril commented:
@MaureenHelm, the conflict was due to #7306 being merged and its changes were left in this PR too. I have removed what was part of #7306 from here, which has removed the conflict.

stanislav-poboril commented:
Hello, since the Zephyr release has been done, could this be merged into the master now? Thanks, S.
